File: get.skewt.new.pl
#!/usr/bin/perl

# This data also available here
#   /home/disk/data/pnw/upperair
# One file per day with all sites concatenated

use LWP::Simple;

$command = "umask 2";
system($command);

#define system env variables
$ENV{'PATH'} = "/usr/local/bin:/usr/ucb:".$ENV{'PATH'};
$ENV{'PYTHONPATH'} = "/home/disk/bob/impacts/bin:/home/disk/shear2/brodzik/python:/usr/local/lib/python2.7/dist-packages:/usr/local/lib/python2.7/lib-tk:".$ENV{'PYTHONPATH'};
$ENV{'LD_LIBRARY_PATH'} = "/opt/intel/compilers_and_libraries_2019.1.144/linux/compiler/lib/intel64_lin";
#$ENV{'DISPLAY'} = ":0";

#$binDir = "/home/disk/bob/impacts/bin";

# Set max height (in km) of plot
$max_ht = 7;

# Go to data directory
#$ARCHIVE_PATH = "/home/disk/user_www/brodzik/olympex/archive/ops";
$ARCHIVE_PATH = "/home/disk/funnel/impacts/archive/ops";
chdir($ARCHIVE_PATH);

# First, get time and make date subdirs if necessary
$yesterday = `date -u --date='1 day ago' "+%Y%m%d"`;
chop $yesterday;
#$yesterday = '20190731';
$yesterdaySkewtDir = $ARCHIVE_PATH."/skewt/".$yesterday;
unless (-e $yesterdaySkewtDir) {
    mkdir $yesterdaySkewtDir;
}
$yesterdayWetbulbDir = $ARCHIVE_PATH."/wet_bulb/".$yesterday;
unless (-e $yesterdayWetbulbDir) {
    mkdir $yesterdayWetbulbDir;
}
$yesterdayTextDir = $ARCHIVE_PATH."/text_sounding/".$yesterday;
unless (-e $yesterdayTextDir) {
    mkdir $yesterdayTextDir;
    $command = "ln -s ".$ARCHIVE_PATH."/text_sounding/index.php ".$yesterdayTextDir;
    system($command);
}
$yesterdayYear=substr($yesterday,0,4);
$yesterdayMonth=substr($yesterday,4,2);
$yesterdayDay=substr($yesterday,6,2);

$today = `date -u "+%Y%m%d"`;
chop $today;
#$today = '20190801';
$todaySkewtDir = $ARCHIVE_PATH."/skewt/".$today;
unless (-e $todaySkewtDir) {
    mkdir $todaySkewtDir;
}
$todayWetbulbDir = $ARCHIVE_PATH."/wet_bulb/".$today;
unless (-e $todayWetbulbDir) {
    mkdir $todayWetbulbDir;
}
$todayTextDir = $ARCHIVE_PATH."/text_sounding/".$today;
unless (-e $todayTextDir) {
    mkdir $todayTextDir;
    $command = "ln -s ".$ARCHIVE_PATH."/text_sounding/index.php ".$todayTextDir;
    system($command);
}

$todayYear=substr($today,0,4);
$todayMonth=substr($today,4,2);
$todayDay=substr($today,6,2);

# Define sites of interest
%siteID = (
    '72518', 'ALB',
    '72528', 'BUF',
    '74494', 'CHH',
    '72208', 'CHS',
    '72632', 'DTX',
    '74389', 'GYX',
    '72403', 'IAD',
    '72426', 'ILN',
    '74560', 'ILX',
    '72501', 'OKX',
    '72520', 'PIT',
    '72318', 'RNK',
    '72402', 'WAL',
    );
@siteNum = (keys %siteID);
$numSites = @siteNum;

# Define skewt times (hours)
#@times = ('00','06','12','18');
@times = ('00','12');
$numTimes = @times;

# Download yesterday's ascii data and create associated skewt
for ($isite=0;$isite<$numSites;$isite++) {

    #print "site = $siteID{$siteNum[$isite]}\n";
    
    for ($itime=0;$itime<$numTimes;$itime++) {

	#print "time = $times[$itime]\n";

	$outFile_skewt = $yesterdaySkewtDir."/ops.skewt.".$yesterday.$times[$itime]."00.".$siteID{$siteNum[$isite]}.".png";
	$outFile_wetbulb = $yesterdayWetbulbDir."/ops.wet_bulb.".$yesterday.$times[$itime]."00.".$siteID{$siteNum[$isite]}.".png";
	$outFile_html = $yesterdayTextDir."/ops.text_sounding.".$yesterday.$times[$itime]."00.".$siteID{$siteNum[$isite]}.".html";
	#print "outFile_html = $outFile_html\n";
	
	# get ascii data
	$urlStr = 'http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR='.$yesterdayYear.'&MONTH='.$yesterdayMonth.'&FROM='.$yesterdayDay.$times[$itime].'&TO='.$yesterdayDay.$times[$itime].'&STNM='.$siteNum[$isite];
	print "urlStr = $urlStr\n";
	$command = "lwp-request '".$urlStr."' > ".$outFile_html;
	system("$command");
	#print "   Done getting yesterdays html file\n'";
	$filesize = -s $outFile_html;
	if ($filesize <= 3000) {
	    print "file too small - go to next time\n";
	    system("/bin/rm $outFile_html");
	}
	else {
	    print "file okay - create skewt\n";
	    #Convert text data to skewt
	    $command = "/home/disk/bob/impacts/bin/removeHtmlTags.csh ".$outFile_html." ".$outFile_html.".txt";
	    system("$command");
	    $command = "/usr/bin/python /home/disk/bob/impacts/bin/removeLinesWithMissingData.py ".$outFile_html.".txt ".$outFile_html.".new";
	    system("$command");
	    $command = "/usr/bin/python /home/disk/bob/impacts/bin/skewplot.py --file ".$outFile_html.".new --outpath /tmp --format UWYO";
	    #print "command = $command\n";
	    system("$command");

	    print "file okay - create T vs Tw plot\n";
	    #Convert text data to T vs Tw plot
	    $command = "/usr/bin/python /home/disk/bob/impacts/bin/vertical_TvsTw_sean_rev_v2.py ".$outFile_html." ".$outFile_wetbulb." ".$max_ht;
	    #print "command = $command\n";
	    system($command);
	    
	    $command = "/bin/rm ".$outFile_html.".txt"." ".$outFile_html.".new";
	    system("$command");
	    $command = "/bin/mv /tmp/upperair.NWS_".$siteID{$siteNum[$isite]}."_sonde.".$yesterday.$times[$itime]."00.skewT.png ".$outFile_skewt;
	    #print "command = $command\n";
	    system("$command");
	    #print "   Done making yesterdays skewt file\n";
	}
    }
}

# Create today images and tables and download each one
for ($isite=0;$isite<$numSites;$isite++) {
    for ($itime=0;$itime<$numTimes;$itime++) {

	$outFile_skewt = $todaySkewtDir."/ops.skewt.".$today.$times[$itime]."00.".$siteID{$siteNum[$isite]}.".png";
	$outFile_wetbulb = $todayWetbulbDir."/ops.wet_bulb.".$today.$times[$itime]."00.".$siteID{$siteNum[$isite]}.".png";
	$outFile_html = $todayTextDir."/ops.text_sounding.".$today.$times[$itime]."00.".$siteID{$siteNum[$isite]}.".html";
	#print "outFile_html = $outFile_html\n";

	# get ascii data
	$urlStr = 'http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR='.$todayYear.'&MONTH='.$todayMonth.'&FROM='.$todayDay.$times[$itime].'&TO='.$todayDay.$times[$itime].'&STNM='.$siteNum[$isite];
	print "urlStr = $urlStr\n";
	$command = "lwp-request '".$urlStr."' > ".$outFile_html;
	system("$command");
	#print "   Done getting todays html file\n";
	$filesize = -s $outFile_html;
	if ($filesize <= 3000) {
	    print "file too small - go to next time\n";
	    system("/bin/rm $outFile_html");
	}
	else {
	    print "file okay - create skewt\n";
	    #Convert text data to skewt
	    $command = "/home/disk/bob/impacts/bin/removeHtmlTags.csh ".$outFile_html." ".$outFile_html.".txt";
	    system("$command");
	    $command = "/usr/bin/python /home/disk/bob/impacts/bin/removeLinesWithMissingData.py ".$outFile_html.".txt ".$outFile_html.".new";
	    system("$command");
	    $command = "/usr/bin/python /home/disk/bob/impacts/bin/skewplot.py --file ".$outFile_html.".new --outpath /tmp --format UWYO";
	    #print "command = $command\n";
	    system("$command");

	    print "file okay - create T vs Tw plot\n";
	    #Convert text data to T vs Tw plot
	    $command = "/usr/bin/python /home/disk/bob/impacts/bin/vertical_TvsTw_sean_rev_v2.py ".$outFile_html." ".$outFile_wetbulb." ".$max_ht;
	    #print "command = $command\n";
	    system($command);
	    
	    $command = "/bin/rm ".$outFile_html.".txt"." ".$outFile_html.".new";
	    system("$command");
	    $command = "/bin/mv /tmp/upperair.NWS_".$siteID{$siteNum[$isite]}."_sonde.".$today.$times[$itime]."00.skewT.png ".$outFile_skewt;	
	    #print "command = $command\n";
	    system("$command");
	    #print "   Done making todays skewt file\n";
	}
    }
}

File: get.skewt.new.py
#!/usr/bin/python

# For some reason skewplot.py has the year hardcoded into it for the NWS soundings
# so this will not run for other years unless that is changed

import os
import sys
import time
from datetime import timedelta
from datetime import datetime
import shutil

# user inputs
debug = 1
bin_dir = '/home/disk/bob/impacts/bin'
catalog_base_dir = "/home/disk/funnel/impacts/archive/ops"
skewt_base_dir = catalog_base_dir+'/skewt'
wetbulb_base_dir = catalog_base_dir+'/wet_bulb'
text_base_dir = catalog_base_dir+'/text_sounding'
max_ht = 7
secsPerDay = 86400
# set startFlag
#    0 -> start now
#    1 -> start at startDate
startFlag = 0
startDate = '20200229'
# set number of days to go back from startDate
num_days = 1
hourList = ['00','03','06','09','12','15','18','21']
siteList = {'72518':'ALB',
	    '72528':'BUF',
            '74494':'CHH',
            '72208':'CHS',
            '72632':'DTX',
            '74455':'DVN',
            '72317':'GSO',
            '74389':'GYX',
            '72403':'IAD',
            '72426':'ILN',
            '74560':'ILX',
            '72305':'MHX',
            '72649':'MPX',
            '72501':'OKX',
            '72520':'PIT',
            '72318':'RNK',
            '72402':'WAL'
}
numHours = len(hourList)
numSites = len(siteList)

# Get start date
if startFlag == 0:
    nowTime = time.gmtime()
    now = datetime(nowTime.tm_year, nowTime.tm_mon, nowTime.tm_mday,
                   nowTime.tm_hour, nowTime.tm_min, nowTime.tm_sec)
else:
    nowDateStr = startDate
    now = datetime.strptime(nowDateStr,'%Y%m%d')

nowDateStr = now.strftime("%Y%m%d")
nowYearStr = now.strftime("%Y")
nowMonthStr = now.strftime("%m")
nowDayStr = now.strftime("%d")

# Make list of dates to process
dateStrList = []
yearStrList = []
monthStrList = []
dayStrList = []
for idate in range(0,num_days):
    deltaSecs = timedelta(0, idate * secsPerDay)
    nextDate = now - deltaSecs
    nextDateStr = nextDate.strftime("%Y%m%d")
    dateStrList.append(nextDateStr)
    nextYearStr = nextDate.strftime("%Y")
    yearStrList.append(nextYearStr)
    nextMonthStr = nextDate.strftime("%m")
    monthStrList.append(nextMonthStr)
    nextDayStr = nextDate.strftime("%d")
    dayStrList.append(nextDayStr)
print >>sys.stderr, 'dateStrList = ',dateStrList
   
# Go to data directory
os.chdir(catalog_base_dir)

for idx,idate in enumerate(dateStrList,0):

    # define year, month and day
    year = yearStrList[idx]
    month = monthStrList[idx]
    day = dayStrList[idx]
    
    # define date dirs and create them if necessary
    skewt_dir = skewt_base_dir+'/'+idate
    if not os.path.exists(skewt_dir):
        os.makedirs(skewt_dir)
    wetbulb_dir = wetbulb_base_dir+'/'+idate
    if not os.path.exists(wetbulb_dir):
        os.makedirs(wetbulb_dir)
    text_dir = text_base_dir+'/'+idate
    if not os.path.exists(text_dir):
        os.makedirs(text_dir)
        command = 'ln -s '+catalog_base_dir+'/text_sounding/index.php '+text_dir
        os.system(command)

    for site in siteList:

        print >>sys.stderr, 'site = ', siteList[site]

        for hour in hourList:

            print >>sys.stderr, 'hour = ', hour

            # define output file names
            outFile_skewt = skewt_dir+'/ops.skewt.'+idate+hour+'00.'+siteList[site]+'.png'
            outFile_wetbulb = wetbulb_dir+'/ops.wet_bulb.'+idate+hour+'00.'+siteList[site]+'.png'
            outFile_html = text_dir+'/ops.text_sounding.'+idate+hour+'00.'+siteList[site]+'.html'
            (base,ext) = os.path.splitext(outFile_html)
            print >>sys.stderr, 'outFile_html = ',outFile_html

            # get ascii data
            try:
                urlStr = 'http://weather.uwyo.edu/cgi-bin/sounding?region=naconf&TYPE=TEXT%3ALIST&YEAR='+year+'&MONTH='+month+'&FROM='+day+hour+'&TO='+day+hour+'&STNM='+site
                command = "lwp-request '"+urlStr+"' > "+outFile_html
                os.system(command)
                print >>sys.stderr, 'Done getting html file'
            except Exception as e:
                print sys.stderr, "lwp-request failed, exception: ", e
                continue

            # check filesize of html file
            fileSize = os.path.getsize(outFile_html)
            if fileSize <= 3000:
                print >>sys.stderr, 'html file too small - go to next hour'
                os.remove(outFile_html)
            else:
                print >>sys.stderr, 'html file okay - create skewt & wetbulb files'
                
                # Remove html tags from file
                command = bin_dir+'/removeHtmlTags.csh '+outFile_html+' '+base+'.txt'
	        os.system(command)
                
                # Remove missing data records from file
	        command = bin_dir+'/removeLinesWithMissingData.py '+base+'.txt '+base+'.new'
	        os.system(command)
                
                # Create skewt and move to field catalog directory
	        command = bin_dir+'/skewplot.py --file '+base+'.new --outpath /tmp --format UWYO --parcel False --hodograph False'
	        os.system(command)
	        command = '/bin/mv /tmp/upperair.NWS_'+siteList[site]+'_sonde.'+idate+hour+'00.skewT.png '+outFile_skewt
	        os.system(command)
                
                # Create wetbulb plot
                command = bin_dir+'/vertical_TvsTw_sean_rev_v2.py '+outFile_html+' '+outFile_wetbulb+' '+str(max_ht)
	        os.system(command)
                
                # Clean up
                command = '/bin/rm '+base+'.txt'+' '+base+'.new'
	        os.system(command)

                print >>sys.stderr, 'Done making skewt & wetbulb files'



